### ISSUE TITLE
Setting an invalid route (even temporarily) borks IP address assignment

### ISSUE THREAD

**Issue by adamierymenko** [zerotier/ZeroTierOne#320]
If you set an invalid route / local network (IP/bits) for a ZeroTier IPv4 managed network, this results in all the members of the network getting borked and requiring a service restart.

Basically the management of OS-level routes and IPs assigned to ZeroTier interfaces needs to be smarter. This overlaps a bit with support for finer grained routing control and work that needs to be done to support default gateway, so we can do it all together. It's not a bug so much as a shortcoming of the current code.

But marking it as a bug because it sure feels like one.

**Comment by kali** [comment]
Could you be more specific about how it manifests ? We are seeing all the post-1.0.3 clients setting a second interface with a wrong ip (ending in a lot of zeroes). A superficial code audit hints at the server thinking it is a network address and the client deciding it's a host address. The IP does not appear anywhere in our admin interface, but it may have been there in the past as a result of a since corrected mishap.

Are we talking about the same thing here ?

**Comment by adamierymenko** [comment]
1.0.3 is a very old version and likely does have such issues... we'd really recommend updating. Latest release is 1.1.4.

This bug refers to a problem in which setting something nonsensical at the controller level like:

"assign IPs in the range 10.1.2.1-10.1.2.254 in the network 10.254.254.0/24" results in devices getting a weird "10.254.254.0" IP address and no "10.1.2._" address. This sticks for a while, though it does eventually seem to sort itself out.

We will be refactoring and cleaning up this code as part of work (already in progress) for #178 so this will get fixed. It's more of a limitation of the current code than a "bug" per se but it certainly feels like a bug if you accidentally bork your config.

**Comment by kali** [comment]
Well, 1.0.3 is the latest version I could find that is actually working. My turn to be more specific:

The clients compiled with master gets this from the server (I call TRACE in setConfiguration)

v4s=10.181.7.0/20,10.181.7.42/20

The second ip is the right one, set statically in the controller ui. The first one is just being weird. x.x.7.0/20 is not a network, the 7's bits are coming after 20. As a consequence, the client "thinks" it's a host ip and just use both. My ifconfig zt0 shows:

```
	inet 10.181.7.0 netmask 0xfffff000 broadcast 10.181.15.255
	inet 10.181.7.42 netmask 0xfffff000 broadcast 10.181.15.255
```

All of the clients in the vpn will exhibit the same double setup, with their right IP and 7.0 as a bonus.

Downgrading the client to 1.0.3 is a workaround that we don't like but allow us to work normally: the server checks the client version is at least 1.0.4 before sending what it thinks are "network routes".

So my understanding is... we have a route somewhere in the database, that the server will perceive as a network route and wants to associate to all the clients for some reason. So the client gets the route, see the unmasked bits are not zeros and decide it's a host route and use it.

**Comment by adamierymenko** [comment]
OH, yes, that's this problem.

The problem is that 10.181.7.0/20 is not a correct network spec. It should be sending:

    v4s:10.181.0.0/20,10.181.7.42/20

That's because /20 gives you a subnet that goes from 10.181.0.0 to 10.181.15.255. 10.181.7.42 is within that subnet.

The current code considers anything that is not a valid network spec with *all zeroes* after the netmask to be an IP address assignment.

**Comment by kali** [comment]
Ok. I think we're in sync.

Now, this broken spec does not appear anywhere in my controller UI — I don't host it, I'm a customer. I guess the inconsistency is persisted somewhere in the database but the UI somehow do not show it, so I can't fix it, and I am stuck with 1.0.3 for now.

You mentioned earlier the problem "seem to sort itself out" eventually. Is there anything me or you can do to help it ? Or should I just wait for the route refactoring to land ?

**Comment by adamierymenko** [comment]
What's your network ID? We can check the DB and possibly fix it there. You can e-mail it to contact@zerotier.com if you don't want to post.

Web UI is getting a rev soon that should hopefully sort some issues out.

**Comment by adamierymenko** [comment]
Closing since this will be addressed with #178 changes.
